packages/stitch/tests/typeMergingWithDirectives.test.ts

```diff
@@ -0,0 +1,499 @@
+    // B. selections from the key can be referenced by using the $ sign and dot notation, so that
+    //    $key.upc refers to the `upc` field of the key.
+    //
+    typeDefs: `
+      ${typeMergingDirectivesTypeDefs}
+      type Query {
+        topProducts(first: Int = 2): [Product]
+        _productsByUpc(upcs: [String!]!): [Product] @merge(argsExpr: "upcs: [[$base.upc]]")
+      }
+      type Product @base(selectionSet: "{ upc }") {
+        upc: String!
+        name: String
+        price: Int
+        weight: Int
+      }
+    `,
+    resolvers: {
+      Query: {
+        topProducts: (_root, args) => products.slice(0, args.first),
+        _productsByUpc: (_root, { upcs }) => upcs.map((upc: any) => products.find(product => product.upc === upc)),
+      }
+    },
+    schemaTransforms: [typeMergingDirectivesValidator],
+  });
+
+  const usernames = [
+    { id: '1', username: '@ada' },
+    { id: '2', username: '@complete' },
+  ];
+
+  const reviews = [
+    {
+      id: '1',
+      authorId: '1',
+      product: { upc: '1' },
+      body: 'Love it!',
+    },
+    {
+      id: '2',
+      authorId: '1',
+      product: { upc: '2' },
+      body: 'Too expensive.',
+    },
+    {
+      id: '3',
+      authorId: '2',
+      product: { upc: '3' },
+      body: 'Could be better.',
+    },
+    {
+      id: '4',
+      authorId: '2',
+      product: { upc: '1' },
+      body: 'Prefer something else.',
+    },
+  ];
+
+  const reviewsSchema = makeExecutableSchema({
+    // @merge directive will direct the gateway to use the tagged resolver to resolve objects
+    // of that type. By default:
+    // 1. the key for that type will be sent to the resolvers first argument.
+    // 2. an array of keys will sent if the resolver returns a list.
+    //
+    // In this example, the `argsExpr` argument for the @merge directive is used to customize
+    // the arguments that the gateway will pass to the resolver.
+    //
+    // This example highlights how using the $ sign without dot notation will pass the entire
+    // key as an object. This allows arbitary nesting of the key input as needed.
+    //
+    typeDefs: `
+      ${typeMergingDirectivesTypeDefs}
+      type Review {
+        id: ID!
+        body: String
+        author: User
+        product: Product
+      }
+      input UserKey {
+        id: ID!
+      }
+      type User @base(selectionSet: "{ id }") {
+        id: ID!
+        username: String
+        numberOfReviews: Int
+        reviews: [Review]
+      }
+      input ProductKey {
+        upc: String!
+      }
+      input ProductInput {
+        keys: [ProductKey!]!
+      }
+      type Product @base(selectionSet: "{ upc }") {
+        upc: String!
+        reviews: [Review]
+      }
+      type Query {
+        _reviews(id: ID!): Review
+        _users(keys: [UserKey!]!): [User] @merge
+        _products(input: ProductInput): [Product]! @merge(argsExpr: "input: { keys: [[$key]] }")
```

Comment:
So what is the string format of argsExp? Is that a parseable GraphQL expression that builds an input?

Reply:
It has to be prepared to turn it into a GraphQL expression. It's not valid GraphQL because the double brackets is not valid. I use that because I want to still allow users to use regular lists with single brackets. The variable format in GraphQL also does not allow periods. After it is prepared or preparsed I sent it through the GraphQL parser so you will get some feedback if it is not valid, although the feedback may be somewhat difficult as we are messages will reflect the the output from my preparation step.

That output is basically that I extract each double brackets and turn it into its own expression and I change the variable names to without periods something like __exp1, __exp2, saving the original names for later use.